File: workspace-engine/backend/app/schemas/page.py
"""Schemas for hierarchical Page resources."""
import uuid
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field


class PageBase(BaseModel):
    title: str = Field(default="Untitled", max_length=255)
    icon: Optional[str] = "📄"
    cover_image: Optional[str] = None
    is_favorite: bool = False
    is_archived: bool = False


class PageCreate(BaseModel):
    workspace_id: uuid.UUID
    parent_id: Optional[uuid.UUID] = None
    title: str = Field(default="Untitled", max_length=255)
    icon: Optional[str] = "📄"
    cover_image: Optional[str] = None
    before_page_id: Optional[uuid.UUID] = None
    after_page_id: Optional[uuid.UUID] = None


class PageUpdate(BaseModel):
    title: Optional[str] = Field(default=None, max_length=255)
    icon: Optional[str] = None
    cover_image: Optional[str] = None
    is_favorite: Optional[bool] = None
    is_archived: Optional[bool] = None
    parent_id: Optional[uuid.UUID] = None
    metadata: Optional[Dict[str, Any]] = None


class PageMoveRequest(BaseModel):
    new_parent_id: Optional[uuid.UUID] = None
    before_page_id: Optional[uuid.UUID] = None
    after_page_id: Optional[uuid.UUID] = None


class PageResponse(PageBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    workspace_id: uuid.UUID
    parent_id: Optional[uuid.UUID] = None
    sort_order: str
    metadata: Dict[str, Any] = Field(default_factory=dict, alias="page_metadata")
    created_at: datetime
    updated_at: datetime


class PageTreeNode(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    workspace_id: uuid.UUID
    parent_id: Optional[uuid.UUID] = None
    title: str
    icon: Optional[str] = "📄"
    is_favorite: bool = False
    is_archived: bool = False
    sort_order: str
    children: List["PageTreeNode"] = Field(default_factory=list)


PageTreeNode.model_rebuild()

File: workspace-engine/frontend/types/page.ts
export interface Page {
  id: string;
  workspace_id: string;
  parent_id: string | null;
  title: string;
  icon: string | null;
  cover_image: string | null;
  is_favorite: boolean;
  is_archived: boolean;
  sort_order: string;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface PageTreeNode {
  id: string;
  workspace_id: string;
  parent_id: string | null;
  title: string;
  icon: string | null;
  is_favorite: boolean;
  is_archived: boolean;
  sort_order: string;
  children: PageTreeNode[];
}

export interface PageCreateInput {
  workspace_id: string;
  parent_id?: string | null;
  title?: string;
  icon?: string;
  cover_image?: string;
  before_page_id?: string;
  after_page_id?: string;
}

export interface PageUpdateInput {
  title?: string;
  icon?: string | null;
  cover_image?: string | null;
  is_favorite?: boolean;
  is_archived?: boolean;
  parent_id?: string | null;
  metadata?: Record<string, unknown>;
}

export interface PageMoveInput {
  new_parent_id?: string | null;
  before_page_id?: string;
  after_page_id?: string;
}
